### ISSUE TITLE
HTTP --max-filesize counts ignored body data

### ISSUE THREAD

**Issue by jay** [curl/curl#14899]
From https://github.com/curl/curl/issues/14440#issuecomment-2346734573 user @MasterInQuestion reports that --max-filesize is applied to the ignored content of a redirect response. For example `curl -v -L --max-filesize 1 http://localhost:8080` and the server's redirect reply has 5 content bytes:

~~~
while true; \
do perl -e 'print ("HTTP/1.1 301\r\nContent-Length: 5\r\nLocation: http://localhost:8080/abc\r\n\r\napple")' | nc -4l localhost 8080; \
done
~~~

curl will fail because 5 bytes of content (even though it's ignored) is more than 1:

~~~
* STATE: DO => DID handle 0x76bd88; line 2164
* STATE: DID => PERFORMING handle 0x76bd88; line 2282
< HTTP/1.1 301
< Content-Length: 5
< Location: http://localhost:8080/abc
* Maximum file size exceeded
~~~

The failure is caused by a check in `Curl_http_size` which AFAICT compares even for body bytes that are not "real" so to speak:

https://github.com/curl/curl/blob/a3bd1dda12ec79cd63e0d81df4ec4b1fbbbcfa1b/lib/http.c#L3273-L3294

There is a separate check that is done in `cw_download_write` which deals with real body bytes, however by that point the download is already taking place.

https://github.com/curl/curl/blob/a3bd1dda12ec79cd63e0d81df4ec4b1fbbbcfa1b/lib/sendf.c#L326-L344

In other words, the former check causes curl to error immediately after the headers are received and the content length is known. The user may want this behavior because if they, say, limit to 1MB then maybe they don't want to download the first 1MB of a 100MB file. OTOH it seems that cw_download_write works in exactly that way, it will allow downloading to the maximum file size.

So if the former check is removed, which is my initial inclination, there will be a change in behavior.

**Comment by MasterInQuestion** [comment]
[ jay @ CE 2024-09-13 15:37:01 UTC:
[ https://github.com/curl/curl/issues/14899#issue-2525121254](https://github.com/curl/curl/issues/14899#issue-2525121254)
    Up: [ https://github.com/curl/curl/issues/14440#issuecomment-2346734573](https://github.com/curl/curl/issues/14440#issuecomment-2346734573)

    @MasterInQuestion reports that:
    ["--max-filesize"](https://curl.se/docs/manpage.html#--max-filesize) is applied to the ignored content, of redirect response.

    E.g.
    `curl -v --max-filesize 1 -L "http://localhost:8080/"`
    ; and the server's redirect reply has 5 content bytes:
[[
```shell
	while true;
	do {
	perl -e '
	$, = "\r\n";
	print (
	"HTTP/1.1 301",
	"Content-Length: 5",
	"Location: http://localhost:8080/abc",
	"",
	"abc", "" );
	' | nc -4l localhost 8080;
	};
	done;
```
]]
    &#x60;curl&#x60; will fail because 5 bytes of content (though ignored) is more than 1:
[[
&#x2A; STATE: DO => DID handle 0x76bd88; line 2164
&#x2A; STATE: DID => PERFORMING handle 0x76bd88; line 2282
< HTTP/1.1 301
< Content-Length: 5
< Location: http://localhost:8080/abc
&#x2A; Maximum file size exceeded
]]

    The failure is caused by a check in "Curl_http_size", which AFAICT:
    Compares even for body bytes that are not "real" so to speak:
    [ https://github.com/curl/curl/blob/4ff04615a095ed1e7c489e3bbd4600b52e68d9a7/lib/http.c#L3273-L3294](https://github.com/curl/curl/blob/4ff04615a095ed1e7c489e3bbd4600b52e68d9a7/lib/http.c#L3273-L3294)
    .
    There is a separate check that is done in "cw_download_write": which deals with real body bytes.
    However by that point the download is already taking place:
    [ https://github.com/curl/curl/blob/4ff04615a095ed1e7c489e3bbd4600b52e68d9a7/lib/sendf.c#L326-L344](https://github.com/curl/curl/blob/4ff04615a095ed1e7c489e3bbd4600b52e68d9a7/lib/sendf.c#L326-L344)
    .
    In other words, the former check causes &#x60;curl&#x60; to error immediately:
    After the headers are received and the content length is known.

    The user may want this behavior because, if they say:
    Limit to 1 MiB, then maybe they don't want to download the first 1 MiB of a 100 MiB file.

    On the other hand, it seems that "cw_download_write" works in exactly that way:
    It will allow downloading to the maximum file size.

    So if the former check is removed, which is my initial inclination:
    There will be a change in behavior. ]
.
    See also: [ https://github.com/curl/curl/issues/14440#issuecomment-2346989774](https://github.com/curl/curl/issues/14440#issuecomment-2346989774)

**Comment by bagder** [comment]
@MasterInQuestion I did not understand one word of this

**Comment by bagder** [comment]
It does not help me. I have no idea what you talk about.  #14900 is my proposed fix.

**Comment by MasterInQuestion** [comment]
I've noted you have tendency to comment before reading... per past events.
    Very strange, as the project's source should be harder to interpret.

    No matter. My proposed design is:
    “Perhaps a separation: "--max-dsize"? (parallel of "fsize")”
    (in ["See also"](https://github.com/curl/curl/issues/14440#issuecomment-2346989774))

**Comment by bagder** [comment]
>   “Perhaps a separation: "--max-dsize"? (parallel of "fsize")”

A separation between what for what purpose? Is that another way of saying my fix is not to your liking?

**Comment by MasterInQuestion** [comment]
Please read "See also":
    “Users of "--max-filesize" may want to limit the overall bytes downloaded:
    Even if it's specifically documented as file size downloaded.
    So I'm not sure that's a bug.”
    (original by jay)

**Comment by bagder** [comment]
I give up.

**Comment by MasterInQuestion** [comment]
Summary:
    |1| New option "--max-dsize": keeping the current behavior (counting all transfer bytes).
    |2| Alias "--max-fsize" to "--max-filesize": changed to count only the final file size.

    2025-09:
    (after the behavior change outright)
    Currently there seems to be no purpose, for the "file" part to present, in the option name at all.
    ["--max-filesize"](https://curl.se/docs/manpage.html#--max-filesize) appears rather frequently used in CLI.
    And the extra length may be a bother to type.
    .
    "--max-size" seems like a good alias.
